concurrent/src/main/scala/scalaz/concurrent/Future.scala

```diff
@@ -235,7 +264,7 @@ sealed abstract class Future[+A] {
 object Future {
   case class Now[+A](a: A) extends Future[A]
   case class Async[+A](onFinish: (A => Trampoline[Unit]) => Unit) extends Future[A]
-  case class Suspend[+A](thunk: () => Future[A]) extends Future[A]
+  case class Suspend[+A](thunk: () => Future[A], andFinally: Option[Option[Throwable] => Future[Unit]]) extends Future[A]
```

Comment:
This algebra makes it impossible to entirely prevent resource leaks.  A more appropriate signature would be `Suspend[+A](thunk: () => Future[A], andFinally: Option[Throwable \/ A => Future[Unit]])`.  As a use case, imagine a resource is acquired asynchronously after preemption; that resource needs to be freed.